koch: add --localdocs to allow building only local docs

alaviss opened the pull request:
This flag also make koch doc use the passed arguments when building
the offline docs.

This is useful when generating nightlies as we would want to use
--doccmd:skip and also skipping a pass of docgen speed things up
drastically (for non-native targets).

This flag superseded the undocumented --docslocal.

alaviss committed:
koch: add --localdocs to allow building only local docs

This flag also make koch doc use the passed arguments when building
the offline docs.

This is useful when generating nightlies as we would want to use
--doccmd:skip and also skipping a pass of docgen speed things up
drastically (for non-native targets).

This flag superseded the undocumented --docslocal.

timotheecour commented on koch.nim:
```diff
@@ -651,19 +658,16 @@ when isMainModule:
-        doAssert op.val.len > 0
-        buildDocsDir(op.cmdLineRest, op.val)
-        break
+      of "local":
```
it's only for docs so you can name it `docslocal` (more self-documenting and discoverable)

alaviss replied:
Opted for `localdocs`, which makes more sense in English.

timotheecour commented on koch.nim:
```diff
@@ -642,7 +646,10 @@ proc showHelp() =
+  var
+    latest = false
+    localOnly = false
+    localOut = ""
```
localOutDir

timotheecour commented on tools/kochdocs.nim:
```diff
@@ -318,6 +318,9 @@ proc buildDocsDir*(args: string, dir: string) =
-proc buildDocs*(args: string) =
-  buildDocsDir(args, webUploadOutput / NimVersion)
-  buildDocsDir("", docHtmlOutput) # no `args` to avoid offline docs containing the 'gaCode'!
+proc buildDocs*(args: string, localOnly = false, localOut = "") =
```
nit: code smell: localOut is only used for localOnly = true; the previous way (when `buildDocsDir` was called directly) was simpler (but i won't insist since this is minor)

timotheecour commented on koch.nim:
```diff
@@ -53,6 +53,10 @@ Options:
+  --localdocs[:path]       only build local documentations. If a path is not
+                           specified (or empty), the default is used.
+                           Additionally options passed to docs will be used to
+                           build local documentation.
```
you could add: `eg: --doccmd:skip` (as was written in comments before this PR) since this is an important use case for localdocs

alaviss replied:
I would rather that we detect and strip `googleAnalytics` from the argument list instead, then you can just reuse the rest of the arguments. ~~But that can be done in a future PR.~~ Or I can just do it now.

alaviss committed:
kochdocs: filter google analytics code from the arg list instead

This commit introduce a small PEG expression to filter out the google
analytics code before building local docs when --localdocs is not
specified. This lets us keep any arguments unrelated to google analytics
when building local docs, useful for use with --doccmd:skip

timotheecour commented on tools/kochdocs.nim:
```diff
@@ -318,6 +318,19 @@ proc buildDocsDir*(args: string, dir: string) =
+  if not localOnly:
+    buildDocsDir(args, webUploadOutput / NimVersion)
+
+    let gaFilter = peg"@( y'--doc.googleAnalytics:' @(\s / $) )"
```
can't we do the opposite, add the `doc.googleAnalytics` flag instead of removing it? seems more robust and cleaner.

for example, in koch.nim processing of `of cmdLongOption, cmdShortOption:` to detect when this flag is passed

alaviss replied:
This code is designed so that you may use your own googleAnalytics code. I'm just trying to preserve old decisions here. Changes to how this work can be done in an another PR.
